zcash_client_sqlite/src/wallet/init/migrations/add_transaction_views.rs

```diff
@@ -0,0 +1,356 @@
+                    stmt_find_utxo_value
+                        .query_row(&[op.hash().to_sql()?, op.n().to_sql()?], |row| {
+                            row.get(0).map(|i| Amount::from_i64(i).unwrap())
+                        })
+                        .map_err(WalletMigrationError::DbError)
+                })?;
+
+                stmt_set_fee.execute(&[i64::from(fee_paid), id_tx])?;
+            }
+        }
+
+        transaction.execute_batch(
+            "CREATE VIEW v_tx_sent AS
+            SELECT transactions.id_tx         AS id_tx,
+                   transactions.block         AS mined_height,
+                   transactions.tx_index      AS tx_index,
+                   transactions.txid          AS txid,
+                   transactions.expiry_height AS expiry_height,
+                   transactions.raw           AS raw,
+                   SUM(sent_notes.value)      AS sent_total,
+                   COUNT(sent_notes.id_note)  AS sent_note_count,
+                   SUM(
+                       CASE
+                           WHEN sent_notes.memo IS NULL THEN 0
+                           WHEN SUBSTR(sent_notes.memo, 0, 2) = X'F6' THEN 0
+                           ELSE 1
+                       END
+                   ) AS memo_count,
+                   blocks.time                AS block_time
+            FROM   transactions
+                   JOIN sent_notes
+                          ON transactions.id_tx = sent_notes.tx
+                   LEFT JOIN blocks
+                          ON transactions.block = blocks.height
+            GROUP BY sent_notes.tx;
+            CREATE VIEW v_tx_received AS
+            SELECT transactions.id_tx            AS id_tx,
+                   transactions.block            AS mined_height,
+                   transactions.tx_index         AS tx_index,
+                   transactions.txid             AS txid,
+                   SUM(received_notes.value)     AS received_total,
+                   COUNT(received_notes.id_note) AS received_note_count,
+                   SUM(
+                       CASE
+                           WHEN received_notes.memo IS NULL THEN 0
+                           WHEN SUBSTR(received_notes.memo, 0, 2) = X'F6' THEN 0
+                           ELSE 1
+                       END
+                   ) AS memo_count,
+                   blocks.time                   AS block_time
+            FROM   transactions
+                   JOIN received_notes
+                          ON transactions.id_tx = received_notes.tx
+                   LEFT JOIN blocks
+                          ON transactions.block = blocks.height
+            GROUP BY received_notes.tx;
+            CREATE VIEW v_transactions AS
+            SELECT id_tx,
+                   mined_height,
+                   tx_index,
+                   txid,
+                   expiry_height,
+                   raw,
+                   SUM(value) + MAX(fee) AS net_value,
```

Comment:
Is the `MAX(fee)` here necessary because this is collating multiple rows with distinct `value` column entries? And in the previous fields, because all collated rows have the same entries you can just select the column and get a single copy of the row?

Reply:
Since we're grouping by the transaction's row ID, we'll get only a single row per transaction (a single row per grouped identifier) so the query cannot even compile without using an aggregate function for each column retrieved from a joined table. `MAX` is correct here because we explicitly set the fee to 0 on the received "side" of the union; when we have two rows with the same transaction ID, we will therefore get the fee amount from the sent-note associated row.